gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/auth/handler/impl/FormLoginHandlerImpl.java

```diff
@@ -74,54 +62,32 @@ public void handle(RoutingContext context) {
             } else if (clientId == null) {
-                log.warn("No client id in form - did you forget to include client_id query parameter ?");
+                logger.warn("No client id in form - did you forget to include client_id query parameter ?");
                 context.fail(400);
             } else {
-                Session session = context.session();
-
                 // build authentication object with ip address and user agent
                 JsonObject authInfo = new JsonObject()
-                        .put("username", username)
-                        .put("password", password)
+                        .put(USERNAME_PARAM_KEY, username)
+                        .put(PASSWORD_PARAM_KEY, password)
                         .put(Claims.ip_address, RequestUtils.remoteAddress(req))
                         .put(Claims.user_agent, RequestUtils.userAgent(req))
                         .put(Parameters.CLIENT_ID, clientId);
 
                 authProvider.authenticate(context, authInfo, res -> {
                     if (res.succeeded()) {
-                        User user = res.result();
-                        context.setUser(user);
-                        if (session != null) {
-                            // the user has upgraded from unauthenticated to authenticated
-                            // session should be upgraded as recommended by owasp
-                            session.regenerateId();
+                        final User result = res.result();
+                        context.getDelegate().setUser(result);
 
-                            // Note : keep returnURLParam in session in case the user go to previous page
-                            // String returnURL = session.remove(returnURLParam);
-                            String returnURL = session.get(returnURLParam);
-                            if (returnURL != null) {
-                                // Now redirect back to the original url
-                                doRedirect(req.response(), returnURL);
-                                return;
-                            }
-                        }
-                        // Either no session or no return url
-                        if (directLoggedInOKURL != null) {
-                            // Redirect to the default logged in OK page - this would occur
-                            // if the user logged in directly at this URL without being redirected here first from another
-                            // url
-                            doRedirect(req.response(), directLoggedInOKURL);
-                        } else {
-                            // Just show a basic page
-                            req.response().end(DEFAULT_DIRECT_LOGGED_IN_OK_PAGE);
-                        }
+                        final String redirectUri = UriBuilderRequest.resolveProxyRequest(context.request(), context.get(CONTEXT_PATH) + "/oauth/authorize?" + req.query());
```

Comment:
I'm not convinced to hard code the `oauth/authorize` value ... 🤔. The same comment apply for the other handlers. 

Login page should not know that it has been called after an OAuth call.  But it means certainly to store this redirectUri into the session.

Reply:
We will use redirect_uri query param do achieve that.